vispy/visuals/tests/test_image.py

```diff
@@ -348,4 +348,44 @@ def test_image_interpolation():
         assert np.allclose(render[center], gray, atol=5)  # we just want gray, this is not quantitative
 
 
+@requires_application()
+def test_image_set_data_different_dtype():
+    size = (80, 80)
+    data = np.array([[0, 127]], dtype=np.int8)
+    left = (40, 10)
+    right = (40, 70)
+    white = (255, 255, 255, 255)
+    black = (0, 0, 0, 255)
+
+    with TestingCanvas(size=size[::-1], bgcolor="w") as c:
+        view = c.central_widget.add_view()
+        view.camera = PanZoomCamera((0, 0, 2, 1))
+        image = Image(data=data, cmap='grays', clim=[0, 127],
+                      parent=view.scene)
+
+        render = c.render()
+        assert np.allclose(render[left], black)
+        assert np.allclose(render[right], white)
+
+        # same data as float should change nothing
+        image.set_data(data.astype(np.float32))
+        render = c.render()
+        assert np.allclose(render[left], black)
+        assert np.allclose(render[right], white)
+
+        # something inverted, different dtype
+        new_data = np.array([[127, 0]], dtype=np.float16)
+        image.set_data(new_data)
+        render = c.render()
+        assert np.allclose(render[left], white)
+        assert np.allclose(render[right], black)
+
+        # out of bounds should wrap around cleanly
+        new_data = np.array([[0, 128]], dtype=np.float64)
+        image.set_data(new_data)
+        render = c.render()
+        assert np.allclose(render[left], black)
+        assert np.allclose(render[right], black)
```

Comment:
Out of bounds in the sense that 128 is over the 127 limit? Shouldn't 128 then be clipped to 127 and made white?

Reply:
No, because:
```py
In [1]: import numpy as np

In [2]: np.array(128, dtype=np.int8)
Out[2]: array(-128, dtype=int8)
```

We can add extra machinery to clip instead, if you think that would be better, hto I'm not sure how that would work. This was intended more as a "do not do crazy stuff like passing the wrong dtype" kind of thing.

Reply:
It's kinda the opposite; the `CPUScaledTexture` *always* sends `np.float32` to the GPU, because the scaling happens on the CPU side. So the flow in the test above is:
- image last time had `int8` data
- `image.set_data()`: data is saved as-is float in `image._data`
- `image._build_texture()`
- `texture.scale_and_set_data()`: here data is converted to the previous `texture._data_dtype` (`int8`) so we can correctly interpret/build clims
- `texture._scale_data_on_cpu()`: converts data to float32 if necessary

Reply:
So in case if we wanted to do the clipping, it should happen in the `scale_and_set_data` step

Reply:
Ah ok, so that casting to the old/original dtype is new behavior for CPU scaling from what I can tell. Lines like:

```
data = np.array(data, dtype=self._data_dtype, copy=copy)
```

didn't exist in the old version from what I can tell in the github diff. I'm ok with that change but it should probably be documented somewhere (in the ImageVisual or VolumeVisual or at least in the texture classes). Based on your comments in the code and your comment above I'm guessing the odd behavior that happens if we don't do this casting is that the limits (and interpolation?) are messed up?

I wasn't suggesting that we add clipping, I think there is clipping in the shader and I thought that's what this test was assuming/using to verify the output. It is actually wrapping/overflowing the floating point 128 64-bit float value to a -128 8-bit signed integer value. That makes sense. I think overall the confusing thing for me was not knowing that the data type is now preserved between `.set_data` calls. Again, that is fine, just not what I expected.

Reply:
Nah, I think any clipping should happen on the GPU/shader. I'd like to avoid any unnecessary processing on the CPU. Especially since this won't apply to most cases where people are using the same data type and they probably have data in the proper range.

Reply:
Saw the answer late, already tried an implementation :P See what you think about it. I agree that ideally you'd do clipping on the GPU, but with a CPU scaled texture we don't really have a choice, it's either we clip it, or the result is gonna be nonsense to the user.